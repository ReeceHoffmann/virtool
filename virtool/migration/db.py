import asyncio
from contextlib import asynccontextmanager


@asynccontextmanager
async def persistent_session(mongo: "Mongo"):
    """Create a persistent mongo db session that won't time out due to inactivity.

    :param mongo: the application MongoDB client
    """
    async with (
        await mongo.client.start_session() as session,
    ):

        async def refresh_session():
            while True:
                await asyncio.sleep(600)
                await mongo.command({"refreshSessions": [session.session_id]})

        refresh_task = asyncio.Task(refresh_session())

        yield session

        refresh_task.cancel()
