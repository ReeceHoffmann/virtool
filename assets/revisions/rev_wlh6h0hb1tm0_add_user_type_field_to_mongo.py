"""add user type field to mongo

Revision ID: wlh6h0hb1tm0
Date: 2024-12-04 16:29:19.348922

"""

import arrow

from virtool.migration import MigrationContext

# Revision identifiers.
name = "add user type field to mongo"
created_at = arrow.get("2024-12-04 16:29:19.348922")
revision_id = "wlh6h0hb1tm0"

alembic_down_revision = "e52b2748f384"
virtool_down_revision = None

# Change this if an Alembic revision is required to run this migration.
required_alembic_revision = None


async def upgrade(ctx: MigrationContext):
    ctx.mongo.users.update_many(
        {"type": {"$exists": False}},
        {"$set": {"type": "user"}},
    )


async def test_upgrade(ctx: MigrationContext, snapshot):
    await ctx.mongo.users.insert_many(
        [
            {"_id": "no type"},
            {"_id": "user_type", "type": "user"},
            {"_id": "bot_type", "type": "bot"},
        ],
    )

    await upgrade(ctx)

    assert await ctx.mongo.users.find({}).to_list() == snapshot()
