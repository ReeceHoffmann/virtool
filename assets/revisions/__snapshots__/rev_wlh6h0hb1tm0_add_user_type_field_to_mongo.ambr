# serializer version: 1
# name: test_upgrade
  list([
    dict({
      '_id': 'no type',
      'type': 'user',
    }),
    dict({
      '_id': 'user_type',
      'type': 'user',
    }),
    dict({
      '_id': 'bot_type',
      'type': 'bot',
    }),
  ])
# ---
